lib/src/util.dart

```diff
@@ -121,6 +121,22 @@ dynamic transformDates(dynamic value) {
   return value;
 }
 
+/// Extract the id from DocumentReference.
+/// Handles complex structures like Map and List
+/// by recursively calling itself on each entry.
+dynamic transformDocumentReference(dynamic value) {
```

Comment:
Instead of copy/pasting this utility, it seems like we could make a reusable method that recursively transforms individual values. How about this:
```dart
/// Transforms a non-recursive value.
typedef TransformSimpleValue = dynamic Function (dynamic simpleValue);

dynamic transformValue(dynamic value, TransformSimpleValue transformSimpleValue) {
if (value is Map<String, dynamic>) {
    return value.map((k, v) => MapEntry(k, transformValue(v, transformSimpleValue)));
  }
  if (value is Iterable) {
    return value.map((e) => transformValue(e, transformSimpleValue)).toList();
  }
  return transformSimpleValue(value);
}
```

Then for your case, you could call it like this:
```dart
isEqualTo = transformValue(isEqualTo, (value) => value is DocumentReference ? value.id : value);
```

And you can refactor the Date transformation in a similar fashion.

Reply:
Thanks for the feedback, code refactored to use common structure and utils.